modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -89,86 +157,468 @@ public IgniteStatisticsRepository statisticsRepository() {
+            // TODO: handle & remove task
+        }
+
+        StatsClearRequest req = new StatsClearRequest(reqId, false, Collections.singletonList(keyMsg));
+
+        sendLocalRequests(reqId, req, requestNodes);*/
+
+        clearObjectStatisticsLocal(keyMsg);
+    }
+
+    /**
+     * Actually clear local object statistics by the given key.
+     *
+     * @param keyMsg Key to clear statistics by.
+     */
+    private void clearObjectStatisticsLocal(StatsKeyMessage keyMsg) {
+        StatsKey key = new StatsKey(keyMsg.schema(), keyMsg.obj());
+        String[] colNames = keyMsg.colNames().toArray(new String[0]);
+
         statsRepos.clearLocalPartitionsStatistics(key, colNames);
         statsRepos.clearLocalStatistics(key, colNames);
         statsRepos.clearGlobalStatistics(key, colNames);
     }
 
+    // TODO
+    /**
+     * Collect local object statistics by primary partitions of specified object.
+     *
+     * @param keyMsg Statistic key message to collect statistics by.
+     * @param partIds Set of partition ids to collect statistics by.
+     * @param cancelled Supplier to check if operation was cancelled.
+     * @return Tuple of Collected local statistics with array of successfully collected partitions.
+     * @throws IgniteCheckedException
+     */
+    private IgniteBiTuple<ObjectStatisticsImpl, int[]> collectLocalObjectStatistics(
+            StatsKeyMessage keyMsg,
+            int[] partIds,
+            Supplier<Boolean> cancelled
+    ) throws IgniteCheckedException {
+        GridH2Table tbl = schemaMgr.dataTable(keyMsg.schema(), keyMsg.obj());
+        if (tbl == null)
+            throw new IgniteCheckedException(String.format("Can't find table %s.%s", keyMsg.schema(), keyMsg.obj()));
+
+        Column[] selectedCols = filterColumns(tbl.getColumns(), keyMsg.colNames());
+
+        Collection<ObjectPartitionStatisticsImpl> partsStats = collectPartitionStatistics(tbl, partIds, selectedCols,
+                cancelled);
+        if (partsStats == null) {
+            assert cancelled.get() : "Error collecting partition level statistics.";
+
+            return null;
+        }
+
+        sendPartStatsToBU(tbl, partsStats);
+
+        StatsKey key = new StatsKey(keyMsg.schema(), keyMsg.obj());
+        statsRepos.saveLocalPartitionsStatistics(key, partsStats);
+
+        ObjectStatisticsImpl tblStats = aggregateLocalStatistics(tbl, selectedCols, partsStats);
+        statsRepos.mergeLocalStatistics(key, tblStats);
+
+        return new IgniteBiTuple(tblStats, partsStats.stream().map(ObjectPartitionStatisticsImpl::partId).toArray());
+    }
+
+    /**
+     * Send statistics propagation messages with partition statistics to all backups node.
+     *
+     * @param tbl Table to which statistics should be send.
+     * @param objStats Collection of partition statistics to send.
+     * @throws IgniteCheckedException In case of errors.
+     */
+    private void sendPartStatsToBU(
```

Comment:
Method 'sendPartitionStatisticsToBackupNodes' name would look more expressive.
I believe method name should has exact meaning, in opposite to variable, that can be short as variable exists in much narrower scope.

Reply:
OK